rflx/pyrflx/typevalue.py

```diff
@@ -260,3 +460,350 @@ def accepted_type(self) -> type:
+            ]
+        ):
+            self._fields[fld].typeval.clear()
+            print([o.condition for o in self._model.outgoing(model.Field(fld))])
+            raise ValueError("value does not fulfill field condition")
+
+        if isinstance(field.typeval, OpaqueValue) and field.typeval.length != field.length.value:
+            flength = field.typeval.length
+            field.typeval.clear()
+            raise ValueError(f"invalid data length: {field.length.value} != {flength}")
+
+        if isinstance(field.typeval, ArrayValue) and field.typeval.length != field.length.value:
+            flength = field.typeval.length
+            field.typeval.clear()
+            raise ValueError(f"invalid data length: {field.length.value} != {flength}")
+
+        self._preset_fields(fld)
+
+    def _preset_fields(self, fld: str) -> None:
+        """
+        Iterates through the following nodes of fld until reaches a node whose successor
+        does not have a first or length field. It sets the first and length fields of all
+        nodes it reaches.
+        """
+
+        nxt = self._next_field(fld)
+        while nxt and nxt != model.FINAL.name:
+            field = self._fields[nxt]
+
+            if not self._has_first(nxt) or not self._has_length(nxt):
+                break
+
+            field.first = self._get_first(nxt)
+            field.length = self._get_length(nxt)
+            if (
+                field.set
+                and isinstance(field.typeval, OpaqueValue)
+                and field.typeval.length != field.length.value
+            ):
+                field.first = UNDEFINED
+                field.length = UNDEFINED
+                field.typeval.clear()
+                break
+
+            self._last_field = nxt
+            nxt = self._next_field(nxt)
+
+    def get(self, fld: str) -> Any:
+        if fld not in self.valid_fields:
+            raise ValueError(f"field {fld} not valid")
+        return self._fields[fld].typeval.value
+
+    @property
+    def to_bitstring(self) -> Bitstring:
+        bits = ""
+        field = self._next_field(model.INITIAL.name)
+        while True:
+            if not field or field == model.FINAL.name:
+                break
+            field_val = self._fields[field]
+            if (
+                not field_val.set
+                or not isinstance(field_val.first, Number)
+                or not field_val.first.value <= len(bits)
+            ):
+                break
+            bits = bits[: field_val.first.value] + str(self._fields[field].typeval.to_bitstring)
+            field = self._next_field(field)
+        if len(bits) % 8:
+            raise ValueError(f"message length must be dividable by 8 ({len(bits)})")
+
+        return Bitstring(bits)
+
+    @property
+    def value(self) -> bytes:
+        bits = str(self.to_bitstring)
+        return b"".join(
+            [int(bits[i : i + 8], 2).to_bytes(1, "big") for i in range(0, len(bits), 8)]
+        )
```

Comment:
`value` is intended to return the Python value of a `TypeValue`, e.g. `int` for `IntegerValue`. Since there is no message type in Python it returns a byte string that is the actual message. This however makes it equal to `to_bytes`. As `value` has to be implemented due to being an abstract property I proposed to remove `to_bytes`.

Reply:
After some separate discussion, saving `value` for #8 and keeping `to_bytes` as is seems like a good option. We only have to think about the behaviour of `value` for messages that are not handled by #8.

Reply:
I think an exception should be raised if the value of a message is not defined as in #8.

Reply:
Fixed by f22a52fa